### ISSUE TITLE
Building Maylan M200 with Marlin 2.0 and PlatformIO

### ISSUE THREAD

**Issue by ghost** [MarlinFirmware/Marlin#11104]
When i try to use the maylan board in config.h, it throws an error "cannot find board id "maylanm200"

**Comment by xC0000005** [comment]
Ok, so to build Marlin for the M200 you need a few more pieces:

From this repo:
https://github.com/xC0000005/M200Toolchain

You get the board .json files (which go in .platformio\platforms\ststm32\boards, and the variant definitions, which go in .platformio\packages\framework-arduinostm32\stm32f1\variants (if you don't see variants under STM32f1, it's probably directly under the arduinostm32 directory).  

Alternately, you can build with STM32GENERIC by doing the following:
Clone from this repo and branch:
https://github.com/xC0000005/Marlin/tree/genericHal
Copy over the M200 configuration files you want to use.
Comment out Serial2 if it's enabled.
In your PlatformIO.ini section for the M200, set platform = https://github.com/xC0000005/platform-ststm32#M200Toolchain

Build as a appropriate. If you build from the generic tree above, it should have board definitions and variants included as it drags down the right builder and so on. 

I haven't done PRs to merge these because:
A. I built the STM32GENERIC package that you're using if you set your platform=
B. I'm not 100% sure these are correct.
C. I'm still waiting for PIO support for ST's official arduino core, and then I'll likely migrate the variants and definitions to it.

That said, you should be able to build by doing so. If your end goal is to enable auto-bed leveling of some sort, there are some hacks you'll need to use to reduce the rom size - you only get 120k to play with after the bootloader and eeporm emulation eats 8k, so whatever you do needs to fit in 112k. It's doable by using dstring & removing the name demangler, or at least it was.

**Comment by ghost** [comment]
Thanks, I will try it tomorrow.  After its built, how does it get on the M200?  Does the M200 have a bootloader?

**Comment by xC0000005** [comment]
For a V1, you copy firmware.bin to update.bin and put it in the root of an SD card along with a file named “fcupdate.flg” (I often copy firmware.bin to BOTH filenames). Put the SD card in the mini, turn it off and back on. If it shows a version of V99.<something> it’s alive. Note the first flash can take longer than the UI knows to wait, so be patient. 

The bootloader is ridiculously picky about what SD card it will flash from - the one that came with it is a good choice, formatted for FAT.

**Comment by ghost** [comment]
Hello,

It now sees the board but I am getting some build errors:

```
Compiling .pioenvs\malyanm200\src\src\HAL\HAL_LPC1768\arduino.cpp.o
In file included from Marlin\src\HAL\HAL_LPC1768\../../inc/MarlinConfig.h:28:0,
from Marlin\src\HAL\HAL_LPC1768\LPC1768_Servo.cpp:63:
Marlin\src\HAL\HAL_LPC1768\../../inc/../HAL/HAL_STM32F1/HAL.h:44:10: fatal error: Arduino.h: No such file or directory

Compiling .pioenvs\malyanm200\src\src\HAL\HAL_LPC1768\include\Wire.cpp.o
In file included from Marlin\src\HAL\HAL_LPC1768\../../inc/MarlinConfig.h:28:0,
from Marlin\src\HAL\HAL_LPC1768\SoftwareSPI.cpp:32:
Marlin\src\HAL\HAL_LPC1768\../../inc/../HAL/HAL_STM32F1/HAL.h:44:10: fatal error: Arduino.h: No such file or directory
```

**Comment by xC0000005** [comment]
What’s your platformio.ini section for the m200 look like?

**Comment by ghost** [comment]
[env:malyanm200]
platform    = ststm32
framework   = arduino
board       = malyanM200
build_flags = !python Marlin/src/HAL/HAL_STM32F1/stm32f1_flag_script.py -DMCU_STM32F103CB -D __STM32F1__=1 -std=c++1y -D MOTHERBOARD="BOARD_MALYAN_M200" -DSERIAL_USB -ffunction-sections -fdata-sections -Wl,--gc-sections
src_filter  = ${common.default_src_filter}
#-<frameworks>
lib_ignore  =
  U8glib
  LiquidCrystal_I2C
  LiquidCrystal
  NewliquidCrystal
  LiquidTWI2
  Adafruit NeoPixel
  TMC2130Stepper
  Servo(STM32F1)
  TMC26XStepper
  U8glib-HAL
  TMC2208Stepper
  c1921b4

**Comment by xC0000005** [comment]
It’s highly likely that PlatformIO has broken variants building once again. I’ll look at the PIO builder tonight, but my guess is that if they’ve done what they usually do, there’s some code that assumes that the variant will have the CPU in the name, and the LD script can be derived from that, and so on (instead of reading from the board’s json file like it used to).

[Builder.zip](https://github.com/MarlinFirmware/Marlin/files/2134559/Builder.zip)

I've attached a zip with a couple of files that overwrite their counterparts under .platformio\platforms\ststm32\builder\frameworks\arduino\ (and one that goes in the subfolder maple, overwriting the F1 builder). It's also a hack but I needed the builder running again. Overwrite your builder files with the right versions from this zip and see if it doesn't come closer.

**Comment by ghost** [comment]
Alot of warnings but now there is only one error:
```
Compiling .pioenvs\malyanm200\src\src\HAL\HAL_STM32F4\HAL.cpp.o

Marlin\src\HAL\HAL_STM32F1\persistent_store_flash.cpp: In function 'void HAL::PersistentStore::read_data(int&, uint8_t*, uint16_t, uint16_t*, bool)':

Marlin\src\HAL\HAL_STM32F1\persistent_store_flash.cpp:96:6: error: ambiguating new declaration of 'void HAL::PersistentStore::read_data(int&, uint8_t*, uint16_t, uint16_t*, bool)'
```

**Comment by xC0000005** [comment]
There’s a PR open to fix this, I believe - 
https://github.com/MarlinFirmware/Marlin/pull/11090/files <https://github.com/MarlinFirmware/Marlin/pull/11090/files>

You can apply those changes and I think it will work (I am using STM32GENERIC, so it’s different).

**Comment by ghost** [comment]
That took care of that but now another popped up:

Marlin\src\module\../inc/../HAL/HAL_STM32F1/HAL.h:126:24: error: '__disable_irq' was not declared in this scope

**Comment by xC0000005** [comment]
Those are ones I’m not familiar with, I’ll look at them soon.

**Comment by ghost** [comment]
Ok, thanks a ton!

**Comment by xC0000005** [comment]
I suspect whatever this is is related to https://github.com/MarlinFirmware/Marlin/pull/10999/files <https://github.com/MarlinFirmware/Marlin/pull/10999/files> - you could try reverting that and seeing if it changes anything.

**Comment by ghost** [comment]
I didn't realize that I had an old copy of Marlin.  I pulled the latest, made the changes to persistent_store_flash.cpp, and it now builds successfully.  I will try to load it onto my M200 tonight.  Thanks.

**Comment by fiveangle** [comment]
https://github.com/MarlinFirmware/Marlin/pull/11120 submitted to add @xC0000005's board defines to the project.

Agreed that #10999 appears to have broken the Maylan M200 build:
```
...
arm-none-eabi-g++ -o .pioenvs/malyanm200/src/src/HAL/HAL_STM32F1/HAL_Servo_Stm32f1.cpp.o -c -fno-rtti -fno-exceptions -std=gnu++11 -std=c++1y -g -Os -ffunction-sections -fdata-sections -Wall -mthumb -nostdlib -mcpu=cortex-m3 -Os -mcpu=cortex-m3 -mthumb -ffreestanding -fsigned-char -fno-move-loop-invariants -fno-strict-aliasing --specs=nano.specs --specs=nosys.specs -MMD -MP -ffunction-sections -fdata-sections --param max-inline-insns-single=500 -march=armv7-m -DF_CPU=72000000L -DMALYANM200 -DPLATFORMIO=30600 -DARDUINO_GENERIC_STM32F103C -DMCU_STM32F103CB -DTARGET_STM32F1 -DMCU_STM32F103CB -D__STM32F1__=1 -DMOTHERBOARD=BOARD_MALYAN_M200 -DSERIAL_USB -DARDUINO=10610 -DBOARD_malyanM200 -DERROR_LED_PORT=GPIOB -DERROR_LED_PIN=1 -DDEBUG_LEVEL=DEBUG_NONE -D__STM32F1__ -DARDUINO_ARCH_STM32F1 -DVECT_TAB_ADDR=134225920 -IMarlin -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/FreeRTOS -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/FreeRTOS/utility -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/EEPROM -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/STM32ADC/src -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/Wire -I.piolibdeps/U8glib-ARM_ID1691/src -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/SPI/src -I.piolibdeps/30aa480 -IMarlin/src/HAL -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/cores/maple -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple/include -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple/usb/stm32f1 -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple/usb/usb_lib -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/variants/malyanM200 Marlin/src/HAL/HAL_STM32F1/HAL_Servo_Stm32f1.cpp
In file included from Marlin/src/HAL/HAL_STM32F1/HAL.cpp:34:0:
Marlin/src/HAL/HAL_STM32F1/HAL.h:43:29: fatal error: libmaple/atomic.h: No such file or directory
#include <libmaple/atomic.h>
^
compilation terminated.
arm-none-eabi-g++ -o .pioenvs/malyanm200/src/src/HAL/HAL_STM32F1/HAL_spi_Stm32f1.cpp.o -c -fno-rtti -fno-exceptions -std=gnu++11 -std=c++1y -g -Os -ffunction-sections -fdata-sections -Wall -mthumb -nostdlib -mcpu=cortex-m3 -Os -mcpu=cortex-m3 -mthumb -ffreestanding -fsigned-char -fno-move-loop-invariants -fno-strict-aliasing --specs=nano.specs --specs=nosys.specs -MMD -MP -ffunction-sections -fdata-sections --param max-inline-insns-single=500 -march=armv7-m -DF_CPU=72000000L -DMALYANM200 -DPLATFORMIO=30600 -DARDUINO_GENERIC_STM32F103C -DMCU_STM32F103CB -DTARGET_STM32F1 -DMCU_STM32F103CB -D__STM32F1__=1 -DMOTHERBOARD=BOARD_MALYAN_M200 -DSERIAL_USB -DARDUINO=10610 -DBOARD_malyanM200 -DERROR_LED_PORT=GPIOB -DERROR_LED_PIN=1 -DDEBUG_LEVEL=DEBUG_NONE -D__STM32F1__ -DARDUINO_ARCH_STM32F1 -DVECT_TAB_ADDR=134225920 -IMarlin -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/FreeRTOS -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/FreeRTOS/utility -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/EEPROM -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/STM32ADC/src -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/Wire -I.piolibdeps/U8glib-ARM_ID1691/src -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/SPI/src -I.piolibdeps/30aa480 -IMarlin/src/HAL -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/cores/maple -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple/include -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple/usb/stm32f1 -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple/usb/usb_lib -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/variants/malyanM200 Marlin/src/HAL/HAL_STM32F1/HAL_spi_Stm32f1.cpp
In file included from Marlin/src/HAL/HAL_STM32F1/../../inc/MarlinConfig.h:28:0,
from Marlin/src/HAL/HAL_STM32F1/HAL_Servo_Stm32f1.cpp:26:
Marlin/src/HAL/HAL_STM32F1/../../inc/../HAL/HAL_STM32F1/HAL.h:43:29: fatal error: libmaple/atomic.h: No such file or directory
#include <libmaple/atomic.h>
^
compilation terminated.
arm-none-eabi-g++ -o .pioenvs/malyanm200/src/src/HAL/HAL_STM32F1/HAL_timers_Stm32f1.cpp.o -c -fno-rtti -fno-exceptions -std=gnu++11 -std=c++1y -g -Os -ffunction-sections -fdata-sections -Wall -mthumb -nostdlib -mcpu=cortex-m3 -Os -mcpu=cortex-m3 -mthumb -ffreestanding -fsigned-char -fno-move-loop-invariants -fno-strict-aliasing --specs=nano.specs --specs=nosys.specs -MMD -MP -ffunction-sections -fdata-sections --param max-inline-insns-single=500 -march=armv7-m -DF_CPU=72000000L -DMALYANM200 -DPLATFORMIO=30600 -DARDUINO_GENERIC_STM32F103C -DMCU_STM32F103CB -DTARGET_STM32F1 -DMCU_STM32F103CB -D__STM32F1__=1 -DMOTHERBOARD=BOARD_MALYAN_M200 -DSERIAL_USB -DARDUINO=10610 -DBOARD_malyanM200 -DERROR_LED_PORT=GPIOB -DERROR_LED_PIN=1 -DDEBUG_LEVEL=DEBUG_NONE -D__STM32F1__ -DARDUINO_ARCH_STM32F1 -DVECT_TAB_ADDR=134225920 -IMarlin -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/FreeRTOS -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/FreeRTOS/utility -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/EEPROM -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/STM32ADC/src -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/Wire -I.piolibdeps/U8glib-ARM_ID1691/src -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/SPI/src -I.piolibdeps/30aa480 -IMarlin/src/HAL -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/cores/maple -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple/include -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple/usb/stm32f1 -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple/usb/usb_lib -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/variants/malyanM200 Marlin/src/HAL/HAL_STM32F1/HAL_timers_Stm32f1.cpp
In file included from Marlin/src/HAL/HAL_STM32F1/HAL_spi_Stm32f1.cpp:39:0:
Marlin/src/HAL/HAL_STM32F1/HAL.h:43:29: fatal error: libmaple/atomic.h: No such file or directory
#include <libmaple/atomic.h>
^
compilation terminated.
arm-none-eabi-g++ -o .pioenvs/malyanm200/src/src/HAL/HAL_STM32F1/persistent_store_flash.cpp.o -c -fno-rtti -fno-exceptions -std=gnu++11 -std=c++1y -g -Os -ffunction-sections -fdata-sections -Wall -mthumb -nostdlib -mcpu=cortex-m3 -Os -mcpu=cortex-m3 -mthumb -ffreestanding -fsigned-char -fno-move-loop-invariants -fno-strict-aliasing --specs=nano.specs --specs=nosys.specs -MMD -MP -ffunction-sections -fdata-sections --param max-inline-insns-single=500 -march=armv7-m -DF_CPU=72000000L -DMALYANM200 -DPLATFORMIO=30600 -DARDUINO_GENERIC_STM32F103C -DMCU_STM32F103CB -DTARGET_STM32F1 -DMCU_STM32F103CB -D__STM32F1__=1 -DMOTHERBOARD=BOARD_MALYAN_M200 -DSERIAL_USB -DARDUINO=10610 -DBOARD_malyanM200 -DERROR_LED_PORT=GPIOB -DERROR_LED_PIN=1 -DDEBUG_LEVEL=DEBUG_NONE -D__STM32F1__ -DARDUINO_ARCH_STM32F1 -DVECT_TAB_ADDR=134225920 -IMarlin -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/FreeRTOS -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/FreeRTOS/utility -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/EEPROM -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/STM32ADC/src -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/Wire -I.piolibdeps/U8glib-ARM_ID1691/src -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/libraries/SPI/src -I.piolibdeps/30aa480 -IMarlin/src/HAL -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/cores/maple -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple/include -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple/usb/stm32f1 -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/system/libmaple/usb/usb_lib -I/Users/speedster/.platformio/packages/framework-arduinoststm32/STM32F1/variants/malyanM200 Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp
In file included from Marlin/src/HAL/HAL_STM32F1/HAL_timers_Stm32f1.cpp:33:0:
Marlin/src/HAL/HAL_STM32F1/HAL.h:43:29: fatal error: libmaple/atomic.h: No such file or directory
#include <libmaple/atomic.h>
^
compilation terminated.
*** [.pioenvs/malyanm200/src/src/HAL/HAL_STM32F1/HAL.cpp.o] Error 1
*** [.pioenvs/malyanm200/src/src/HAL/HAL_STM32F1/HAL_Servo_Stm32f1.cpp.o] Error 1
*** [.pioenvs/malyanm200/src/src/HAL/HAL_STM32F1/HAL_spi_Stm32f1.cpp.o] Error 1
*** [.pioenvs/malyanm200/src/src/HAL/HAL_STM32F1/HAL_timers_Stm32f1.cpp.o] Error 1
In file included from Marlin/src/HAL/HAL_STM32F1/../../inc/MarlinConfig.h:28:0,
from Marlin/src/HAL/HAL_STM32F1/persistent_store_flash.cpp:32:
Marlin/src/HAL/HAL_STM32F1/../../inc/../HAL/HAL_STM32F1/HAL.h:43:29: fatal error: libmaple/atomic.h: No such file or directory
#include <libmaple/atomic.h>
^
compilation terminated.
*** [.pioenvs/malyanm200/src/src/HAL/HAL_STM32F1/persistent_store_flash.cpp.o] Error 1
========================== [ERROR] Took 4.49 seconds ==========================
```

**Comment by xC0000005** [comment]
This is more likely to be PIO screwing up variant builds again. There's a rework in progress of the builders and they make some bad assumptions about how to find variant directories, names, include builders, set linker scripts, vector tables, and a few other things.

**Comment by xC0000005** [comment]
Ah, I believe I've figured it out. In the board.json file, it's now required to have a .variant property under the build section which matches the CPU variant:
Currently, it's 
"variant": "malyanM200",
But PIO's newest builder expects:
"variant": "stm32f103cb" to invoke the STM32F1 builder.
It could be that this is an error, and the intent is actually to switch on build.mcu_type, but that change alone would not fix the problem with stm32F1.py, which doesn't handle variants with specific names, ld scripts, and so on.

**Comment by ghost** [comment]
Thanks again for all your help.  Much appreciated.  I loaded the firmware.  Everything works but the fan control.  Do have to change something somewhere?

**Comment by xC0000005** [comment]
The M200 config has the fan set to auto - does it not turn on when the temp is > 50 degrees celsius?

**Comment by ghost** [comment]
It does not.  It was working with the original firmware right before I loaded Marlin.  I will check the hardware tomorrow and see if its trying to turn the fan on.  Maybe something happened to my board or fan but doubtful.

Also, can I disable the auto and have GCode control?  I have a hardwired hotend cooling fan to 12v and I also have a layer fan on the mainboard fan plug.   I want the layer fan to be controllable through Gcode.

**Comment by xC0000005** [comment]
Absolutely - in advanced configuration.h look for PB8, which is currently assigned to E0_AUTO_FAN somthing. Comment that out and set FAN_PIN in pins.h to PB8 (it’s commented out). The sample configs are from my printer where fan 1 (the default one) is the auto fan) and fan 2 (the unpopulated header on V1) is the layer fan.  I’m not sure what would be stopping it, to be honest, but you can check and see if manual control works. Also, if the LCD shows the right temps, the ADC is reading the data correctly, so it must be something about the config.

**Comment by ghost** [comment]
So here was the problem.  #define E0_AUTO_FAN_PIN -1.  I didn't start with your config apparently.
I changed the Maylan pin.h.  It turns on and off with GCode now but PWM isnt working.  Its either 0 or 100% no matter what speed I try to set it to (ex M106 P0 S125 - 50%).

**Comment by ghost** [comment]
```cpp
#ifndef FAN_PIN
  #define FAN_PIN        PB8   // FAN1 header on board - PRINT FAN
#endif
#define FAN1_PIN           PB3   // FAN2 header on board - CONTROLLER FAN
#define FAN2_PIN           -1    // FAN3 header on board - EXTRUDER0 FAN
```

**Comment by xC0000005** [comment]
Enable soft pwm in the configuration.

**Comment by xC0000005** [comment]
#define FAN_SOFT_PWM in configuration.h should give you fan speed control

**Comment by ghost** [comment]
Trying this now.  Is your config in the examples folder?  If not, can you send it to me?  I want to compare settings.  Thanks

**Comment by xC0000005** [comment]
Its the one in the examples folder.

**Comment by ghost** [comment]
Thanks, that worked for PWM.  I never had to change any of that before.  Wondering what config files I started with.

**Comment by xC0000005** [comment]
The M200 doesn’t have hardware PWM, and even if it did the fan isn’t hooked to the right pins for it. Make sure steps per mm is right for your variant of the M200 and the endstops are inverted correctly for your setup.

**Comment by xC0000005** [comment]
That reply made no sense. It doesn’t have hardware PWM set up. The fan pins are PB3 and PB8, so software is the only choice (I hope that makes more sense.)

**Comment by ghost** [comment]
Ok, that makes more sense.  I was going to say that I thought the STM32F103 had hardware PWM.  Everything looks good so far.  I am going to try printing tonight so I will see how that goes.

**Comment by ghost** [comment]
I am having issues with the LCD.  It is very flaky.  It doesn't always update the correctly and it gets stuck alot.  For example, I am running a simple GCODE file for bed leveling that goes to each corner and when it completes the progress bar is stuck at 94%.  Also, if I change  either temp while its in print mode it locks up.  Do you think I need to update the LCD?  I have never done that.  Its a V1, if I need to update it can you send me the LCD update?

**Comment by xC0000005** [comment]
The LCD code is pretty much my own. I will see if I can figure out what keeps the completion screen from popping. The LCD is picky about getting exactly the right messages in the right order. Temperature wise I haven’t seen the printer lock up but I will see if the right command is being sent. If it enqueues a wait, it would wait.

**Comment by ghost** [comment]
Thanks.  One more thing, can I specify how many characters I want to show up of the filenames on the SD card. The original firmware showed alot more than Marlin does.

**Comment by fiveangle** [comment]
`#define SCROLL_LONG_FILENAMES`

**Comment by xC0000005** [comment]
That, too, is all my code. I will see.

**Comment by xC0000005** [comment]
Try this - in malyanlcd.cpp, line 345, change card.filename to:
card.longFilename[0] ? card.longFilename : card.filename

and see if it shows the long filenames.

**Comment by xC0000005** [comment]
Ok, I know the fix for the long file names. I’ll create a PR. Both instances of card.filename need to be switched, and it works fine.

**Comment by fiveangle** [comment]
If the board json needs updating too, please do the needful

**Comment by ghost** [comment]
My line 345 in maylanlcd.cpp does not contain card.filename.  My line 345 is "write_to_lcd_P(PSTR("{SYS:OK}"));"

![2018-06-28 07_33_23-malyanlcd cpp c__users_jason_desktop_m200_marlin-bugfix-2 0 x--6-26-18 atom](https://user-images.githubusercontent.com/21220408/42031806-95881856-7aa5-11e8-8830-89b8fa16313a.png)

**Comment by ghost** [comment]
I have also noticed that it takes a very long time to cancel a print as well.  Any thoughts on the buggyness of the LCD?

**Comment by xC0000005** [comment]
I’m waiting on answers as to how to lay out the variant and adjust the names. Once I have that, I’ll renamed everything and make the changes, but stm32duino and the new ST arduino core are both having similar issues in not working with variants the same way they once did. stm32generic does not, but that’s because I modified the builder in my repo to cope with it. Once PIO settles on a solution, I’ll use it.

**Comment by xC0000005** [comment]
First off, let me settle some terms so we’re talking about the same thing. The LCD UI (running on an ESP8266 MCU) is not the problem here. Malyan’s code is directly integrated with the LCD reader, parsing commands over USART1 and directly mucking with the printer’s internal objects in response to commands.

When I set out to build an implementation for Marlin, I wanted to make it work with as few changes as possible to the core of the code. I knew from watching the debug output from Malyan’s motion controller code (which you can do if you hook a USB->TTL converter to the display port at 500,000, 7, N, 1) that some of their operations used enqueue_and_echo. Why do I know it? Because it enqueues and echos that it’s doing so.  

So my solution was to make all the operations inject gcode using the enqueue and echo call. Gcode becomes the “interface” by which the LCD affects changes, if that makes sense, BUT there are some downsides to this approach. 
For instance, if a gcode file sets temp using the “wait” parameter and you adjust it via the LCD, this just doesn’t work. Why? Because the LCD code enqueues a temperature change, but that won’t be executed until after the current one completes. 

I may actually change that to not use enqueue specifically for that reason. 
Cancelling a print is literally no different than the cancel from other LCDs. I mean that because I copied the cancel code at the time I wrote it. (I should have refactored to make a “cancel print” function.)

Anyway, the buggyness of the LCD is pretty much directly related to me implementing the functionality in a way that impacted Marlin’s core code in the least way possible. I do think it makes sense to move to using the thermal manager’s setTargetHotend function to avoid the delay with setting temperatures, and I’m 100% aware that feedrate support doesn’t work.  Long file names I have a fix for, so it’s slowly getting better.

**Comment by ghost** [comment]
First, thanks for all your work to get Maylan working.  My initial thought was to replace the controller and screen with a Ramps/Graphic LCD combo.  I was pleasantly surprised someone already had this working with Marlin so I didn't need to do all that.

That makes sense.  That explains some of what I am seeing on the UI.  From a UX standpoint, directly hooking into the core of Marlin would be alot better instead of enqueueing commands.  I always thought the Marlin LCD cancel has been kinda annoying but useable. 

I am a Web/.Net coder and have a lot of electronics experience.  I don't really get deep into C/C++ but please let me know if there is anything I can do to help.

**Comment by xC0000005** [comment]
[malyanlcd.cpp.zip](https://github.com/MarlinFirmware/Marlin/files/2146464/malyanlcd.cpp.zip)
This malyanlcd.cpp has the long file name and target temp fixes (note I've only had time to test LFNs). See if it behaves better. Print cancel is no different, feedrate still doesn't work (and isn't even documented in their MMS wiki). I'll fix feedrate next if we can get temp and filenames behaving correctly.

**Comment by ghost** [comment]
I will try it later tonight and let you know how it works.  Any plans to change the Cancel?

**Comment by xC0000005** [comment]
Not at the moment. It appears to be doing the same thing other cancels are, so I’ll need to look at how it’s behaving and why. My goal is to get all LCD functions working at some level, then look at improvements.

**Comment by ghost** [comment]
Long filenames are fixed.  I cannot verify the temps.  It immediately is going to the print finished screen when clicking on a file.

**Comment by ghost** [comment]
I commented out line 416 `//write_to_lcd_P(PSTR("{TQ:100}"));` to get it to print.  Good news is that it looks like the Temp is fixed.  I can change it while it's heating up and while it's printing.  The display echos back the correct temp as well.

**Comment by ghost** [comment]
I think I see the issue with the finished.  card.sdprinting is one of the flags used to signal that a print has finished for the UI.  card.openAndPrintFile is used to start the print but it doesn't set sdprinting to true though so its always false.

**Comment by xC0000005** [comment]
Yes - that’s probably what’s wrong. I’m working on getting my printer back in one piece so I can look at it (I like the M200 but the bed wire design is terrible).

**Comment by xC0000005** [comment]
OpenAndPrintFile enqueues an M23, and M23 calls card.startFilePrint(), which will set it to true, but this allows for a window that I think I understand. Once I test something, I'll create a PR.

**Comment by xC0000005** [comment]
I understand a great deal more about what's going on. There are some subtle problems and some not so subtle ones.
First off, the display code is way, way too chatty. It's emulating the stock MC firmware but the current code sends way too much data. Secondly, there's a command missing after the {TQ:100}, I'm guessing it's either {SYS:DONE} or {SYS:OK}, but after sending the {TQ:100} it needs one of those (I'm out of town until Monday but then I'll listen in on a stock MC firmware print and figure out which one it is). I've reduced the updates to be only when the print percentage changes, or the UI asks for temperature/percent status, and that cuts out 99% of the serial traffic.

**Comment by ghost** [comment]
Can you send your latest code and I will test?  Thanks

**Comment by xC0000005** [comment]
https://raw.githubusercontent.com/xC0000005/Marlin/genericHal/Marlin/src/lcd/malyanlcd.cpp <https://raw.githubusercontent.com/xC0000005/Marlin/genericHal/Marlin/src/lcd/malyanlcd.cpp>

Should be functional except for not showing the “print completed” screen at 100%.

**Comment by ghost** [comment]
This is alot better.  It seems to fix almost all of the issues I was seeing.  It did show the print finished as well for the first print but not the second.  The second one went right back to the file menu but it actually started the print.  When I went to start another print after that, it stopped printing completely.  Also, before, moving any of the axis was very buggy but now it seems really smooth and responsive.  I will use it more and let you know if I see any more issues.

Should setting the speed work?

**Comment by thinkyhead** [comment]
> Should setting the speed work?

If you mean, adjusting the feedrate during a print, then yes, it should work, as long as it's still under the maximum speed and acceleration limits.

**Comment by xC0000005** [comment]
I don’t believe I implemented the feedrate menu command yet. I will fix that soon once I get print completion showing consistently. (Wow, thanks, autocorrect for autocorrecting to absolute garbage).

**Comment by ghost** [comment]
After using it a bit more there are other issues.  Sometimes the temps do not display.  They stay at zero.

**Comment by xC0000005** [comment]
Does the printer heat up and function otherwise?  I suspect I may need a maximum update time to force a refresh. If it isn’t heating that’s one problem. If the LCD isn’t being told of temp changes that’s another.

**Comment by xC0000005** [comment]
I’ve updated the code on github (same location) with support for feedrate, but I’ll need to add diagnostics for the zero temp issue - I can’t repro it if I manually issue a status comand, it always sends the temp status, and {B:0} is resulting in temperature as well. (If you preheat, the LCD starts an S:I loop that continues, and if you don’t, it’s B:0). Can you tell me how you are hitting the zero temp readings?

**Comment by Injineers** [comment]
Ill check my printer out in a bit as well as I have been running it on my controller a bit, though I have only ran a handful of prints so far, 

only issues I have are those that we discussed in the other thread

**Comment by ghost** [comment]
I cant seem to reproduce the temps going to 0.  I am using it now.  The only thing I noticed so far was that the SD card is a little buggy.  Sometimes it shows the files and sometimes is says "Please wait" or "SD Card Empty".  I will use it more and report back.  Thanks

**Comment by jannau** [comment]
The include `<libmaple/atomic.h>` error is caused by [moving the header in Arduino_STM32](https://github.com/rogerclarkmelbourne/Arduino_STM32/commit/3f7db854e4072c98fe435126e49f52e62e816256). With the include fixed to `<util/atomic.h>` compiling works but linking fails with "`cannot open linker script file stm32f103.ld: No such file or directory`".

*edit* linking fixed by changing `platforms/ststm32/builder/frameworks/arduino/maple/stm32f1.py` to use `ldscript = "jtagOffset.ld"`. No success on the M100/Fabrikator 2 Mini though.

**Comment by vsavic** [comment]
@jannau did you change stm32f1.py which comes with installation, or version from here: https://github.com/MarlinFirmware/Marlin/issues/11104#issuecomment-400078289

Can you post your modified file?

**Comment by jannau** [comment]
@vsavic I copied the [variant directory](https://github.com/xC0000005/M200Toolchain/tree/master/STM32duinoVariantDefintion/malyanM200) to packages/framework-arduinoststm32/STM32F1/variants/ and used following diff against the python files [m200_pio_diff.txt](https://github.com/MarlinFirmware/Marlin/files/2307075/m200_pio_diff.txt) using platform-ststm32 4.4.0
It builds fine but does't appear to work with Malyan M100/Fabrikator 2 Mini. I can't tell if that's due to build issues or subtle differences between the M100 and M200. Malyan uses the same firmware for both.
Please note: I don't know what I'm doing, I'm just fixing the issues I saw.

**Comment by xC0000005** [comment]
What does it do with the M100? I’ve seen an M100  running the M200 display, so I’d guess they’re near identical.

**Comment by xC0000005** [comment]
@jannau , I checked with someone who knows the board revisions fairly well. Is your M100 STM32F070 based? Many are though I have pictures of one with a 103 MCU - same with the delta, there's at least one 103 revision, but most (if not all production released ones) are 070.

**Comment by jannau** [comment]
@xC0000005 mine is STM32F103C8 based (first Fabrikator 2 Mini batch) and my understanding is that it is electronically identical except for the missing display.

**Comment by xC0000005** [comment]
It should be. What happens when you reflash with Marlin? Does it not move? Not init USB?

**Comment by jannau** [comment]
With my Marlin build flashed the stepper motors are enabled and the single multi color led lights up blue and that's it. USB does't seem to be initialized. I never enabled/used the web-interface.

**Comment by xC0000005** [comment]
Where does the LED plug in?  Or is it built into the main board?

**Comment by jannau** [comment]
It's on the main board integrated in a push button. Above the uSD slot and the usB connector on the right side in this photo 
![img_20170520_215555](https://user-images.githubusercontent.com/486189/44422306-123bae00-a583-11e8-9daa-921e60fa6940.jpg)

**Comment by xC0000005** [comment]
Ok. I have zero idea if that will work. I’d suggest you build the M200 diagnostics and start testing to see if the pins are the same. The M200 diagnostics don’t initialize much and will let you see if the fan, heater, etc pins are the same.

https://github.com/xC0000005/M200Diagnostics <https://github.com/xC0000005/M200Diagnostics>

The resulting bin (renamed to update.bin) should initialize USB and you can connect via a serial monitor and poke at the hardware.

**Comment by jannau** [comment]
No difference with M200Diagnostics. No USB and stepper motors stay enabled. I don't think the problem is caused by different PINs as the single v29 motion firmware from the malyan wiki works on my M100.

**Comment by xC0000005** [comment]
Sorry, I'm not sure what is happening. I'll keep an eye out on Ebay and if I can find a cheap M100 main board I'll see if I can figure it out. If you really, really want to dig into it, I have a function called "turnOnFan" and I start by modifying the startup assembly to branch to it right away. If the fan comes on...I know my code is at least running. :)  The Diagnostics firmware starts by turning everything off, so I suspect it's not ever initializing. I can see in the picture that there's still a USB pullup transistor that has to be toggled for enumeration (it's not present on the V2 or the new M100 version mainboards), and it's likely to still be PB9, but we're down to stepping our way through init to figure out where things go wrong. This board could actually have ERROR_LED defined correctly, which would make some diagnostics easier.  Where does the ribbon cable off to the side lead?

**Comment by jannau** [comment]
The ribbon cable connects Y stepper and end stop, fan and hotend thermistor and heater (questionable design choice). I also think that the diagnostics firmware is not initialized. I stripped it down to just disable everything on startup without success. So that means either there's a problem in my toolchain or something in the M200 variant directory trips on my HW.

**Comment by thinkyhead** [comment]
@jannau — For some reason I can't get the `malyanm200` env building due to the same old error. The compiler doesn't find `util/atomic.h` or `libmaple/atomic.h` in the platform's `include` folder, even though both exist and are identical. Is there some trick that I'm forgetting to apply?

**Comment by xC0000005** [comment]
Did you apply the builder patch I attached? I got confirmation recently I’m going to need to do a PR to fix up the board names and linker scripts names inside the builder scripts, and also that I’ll probably have to rename the variants to include their MCU types in the names.

**Comment by jannau** [comment]
@thinkyhead I did modify ststm32 4.4.0 in .platformio as described [here](https://github.com/MarlinFirmware/Marlin/issues/11104#issuecomment-414732034). It compiles and links successfully but I'm not convinced it produces working binaries.

**Comment by xC0000005** [comment]
@jannau , if you want to mail your firmware.bin file to me (git hub handle AT gmail dot com) I'll load it on to my printer and test. I have most revisions of the malyan mainboards (except the M100 one).

**Comment by jannau** [comment]
@xC0000005 thanks, email sent. sha256sum: 2168e3a87516b65da3b101fe9e9e0bbf73f4325d16dbd6dca1e6d8785048a425  jannau_2018-08-22_malyanm200_firmware.bin

**Comment by thinkyhead** [comment]
I'm looking for a solution that doesn't involve modifying the platform files.

**Comment by xC0000005** [comment]
Per https://github.com/platformio/platform-ststm32/issues/127 <https://github.com/platformio/platform-ststm32/issues/127>, It’s expected to do some fixup in the platform builder. Once ST’s official core is supported by PIO, I will create official variants for all three boards in that framework, and we won’t need these fixups (mostly - there’s some oddities in ST’s startup code).

**Comment by xC0000005** [comment]
Your firmware is working, for some degree of working. It initializes USB, I was able to connect and issue commands, but steps per mm were wrong, axis direction and endstops were wrong, and the malyan LCD wasn’t enabled. So your build does “work” but it looks like the configuration options are wrong.

My best guess is that the configuration.h and configuration_adv.h from the M200 directory weren’t copied to the right place (???).

**Comment by jannau** [comment]
Thanks for the test. No idea what happened with configuration.h and configuration_adv.h. They should be correct ones. The Marlin configuration should be unrelated the problems I see. So there seems to be a difference between the M100 and M200 which currently prevents the same firmware image from running.

**Comment by vsavic** [comment]
Just wanted to share one more info about what I had to do in order to build this Marlin 2.0 version, and how to fix this error:

> error: 'F_CPU' was not declared in this  #scope

So in `STSTM32\builder\main.py` there's been some changes. 

When I used this version, I managed to build it for `M200`. [Here's](https://github.com/platformio/platform-ststm32/commit/772ece542c6075a03cabaa347a637482c20c207c#diff-7bf8a7e6a6e69b354a3d3e106cc0cbac) how the file looked before these changes.



My guess is that maybe most of you who managed to compile this version without additional modifications had this STSTM32 installed earlier, and you didn't update it maybe for some time.

**Comment by Injineers** [comment]
hmm......

im at a loss, theres alot of threads pointed all around on how to fix this, but I cant seem to get it working

I am getting this 
fatal error: libmaple/atomic.h

I tried updating it to the other location 'util/latomic.h

but got an error that it did not exist

I then DL'd  xC0000005 builder.zip & replaced the files & still got fata error on 'util/latomic.h I then swtich the Hal.h back to libmaple/atomic.h but am still getting that it doesnt exists

**Comment by vsavic** [comment]
Did you made a typo here maybe: `util/latomic.h` vs `util/atomic.h` ?

**Comment by Injineers** [comment]
nah still same issue after removing the l before atomic

<details><summary>#Click to see Pio Error</summary>

```
Processing malyanm200 (platform: https://github.com/xC0000005/platform-ststm32#M200Toolchain; board: malyanM200; framework: arduino)
 
Verbose mode can be enabled via `-v, --verbose` option
PLATFORM: ST STM32 > Malyan STM32F103CB (20k RAM. 128k Flash)
SYSTEM: STM32F103CB 72MHz 20KB RAM (128KB Flash)
DEBUG: CURRENT(stlink-v2) EXTERNAL(blackmagic, stlink, stlink-v2, stlink-v2-1)

Converting Marlin.ino
Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF MODES: FINDER(chain) COMPATIBILITY(soft)
Collected 3 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <Arduino-L6470> 0.6.2
|-- <30aa480>
Compiling .pioenvs\malyanm200\src\src\HAL\HAL_STM32F1\HAL.cpp.o
Compiling .pioenvs\malyanm200\src\src\HAL\HAL_STM32F1\HAL_Servo_Stm32f1.cpp.o
Compiling .pioenvs\malyanm200\src\src\HAL\HAL_STM32F1\HAL_spi_Stm32f1.cpp.o
Compiling .pioenvs\malyanm200\src\src\HAL\HAL_STM32F1\HAL_timers_Stm32f1.cpp.o
Compiling .pioenvs\malyanm200\src\src\HAL\HAL_STM32F1\persistent_store_flash.cpp.o
Compiling .pioenvs\malyanm200\src\src\HAL\HAL_STM32F1\persistent_store_impl.cpp.o
Compiling .pioenvs\malyanm200\src\src\HAL\HAL_STM32F1\persistent_store_sdcard.cpp.o
In file included from Marlin\src\HAL\HAL_STM32F1\HAL.cpp:34:0:
Marlin\src\HAL\HAL_STM32F1\HAL.h:43:10: fatal error: util/atomic.h: No such file or directory
```
</details>

**Comment by vsavic** [comment]
If you are on windows, try also to remove this: `C:\Users\replace.this.with.your.user\.platformio\packages\framework-arduinoststm32` and run a build again after you changed builder in `platforms`. 

EDIT
And after that check here to see if you have util folder with atomio.h file in it or it's in libmaple:
`C:\Users\**your.user**\.platformio\packages\framework-arduinoststm32\STM32F1\system\libmaple\include\`

**Comment by Injineers** [comment]
Still the same.

I think I am going to purge everything & start over

**Comment by jnoxon** [comment]
This thread is amazing. An hour ago I had never heard of PlatformIO and I now have an M200 build from the 2.x branch. It's getting late but I am definitely going to try it out tomorrow!

**Comment by jnoxon** [comment]
Apparently my ability to get it to build does not extend to getting it to work on the printer. Something gets flashed (no version is displayed) and then the motion controller is dead afterward. I was able to recover with the v29 Malyan firmware, and tried flashing it again, and got the same result. I used the STM32f1 variant. Any suggestions would be greatly appreciated!

**Comment by xC0000005** [comment]
Did you copy over the config files with the M200 ones?

**Comment by jnoxon** [comment]
@xC0000005 I have to admit I forgot to do that 🤦‍♂️, so I tried a clean build with the right configs and looked at the binary through strings and it seems okay, but the machine is not acting any differently.

**Comment by xC0000005** [comment]
Does it appear in usb?  It will be called “Maple.”

**Comment by jnoxon** [comment]
It doesn't appear in USB. With the stock firmware I see this:

```Bus 020 Device 020: ID 2974:0503 2974 Malyan 3D Printer Port  Serial: 6D8721A9xxxx```

It might also be worth mentioning that something is making buzzing noises (steppers?) that completely goes away once the stock firmware boots.

**Comment by xC0000005** [comment]
I’m about as far as someone could be from their 3d printers (literally on a beach in the Bahamas), but do this:
Run pio -v -e malyanm200, and mail the entire output plus the fimrware.bin file to my githubhandle (at) gmail dot com.  I’ve got a hex editor handy and can load up the firmware in binary ninja to see if it looks even remotely right.

**Comment by Alfrederson** [comment]
> It doesn't appear in USB. With the stock firmware I see this:
> 
> `Bus 020 Device 020: ID 2974:0503 2974 Malyan 3D Printer Port Serial: 6D8721A9xxxx`
> 
> It might also be worth mentioning that something is making buzzing noises (steppers?) that completely goes away once the stock firmware boots.

I could get it to compile right away selecting the STM32F1 environment, but since the MCU is different (STM32F103RE) the board will do nothing even if it can be flashed with it. If I change the MCU to STM32F103C8, then I get the same errors @fiveangle had and can't get it to compile.

**Comment by Injineers** [comment]
This is still broken

I purged everything, used latest stm from platformio (since not it seems to include the malyanm200) atomic.h error again.

I then removed the latest stm & installed 4.5 same issue.

I then went back to this thread & grabbed the builder files, variants, etc...still errors.

anyone have any idea? or is it possible we can get a updated file structure for those who may be installing atom & platformio new & using the latest

**Comment by kajsilander78** [comment]
> This is still broken
> 
> I purged everything, used latest stm from platformio (since not it seems to include the malyanm200) atomic.h error again.
> 
> I then removed the latest stm & installed 4.5 same issue.
> 
> I then went back to this thread & grabbed the builder files, variants, etc...still errors.
> 
> anyone have any idea? or is it possible we can get a updated file structure for those who may be installing atom & platformio new & using the latest

Try building with stm32duino 1.5 in arduino IDE, if you know how to. I have not tried it myself yet. Tried with early versions and got it to compile, but had some issues with x&y movement, z worked fine but moved wrong direction. Will test again when I get the time.... If any success I'll update...

Here is a guide to start with:
1) https://github.com/stm32duino/wiki/wiki/Getting-Started
In arduino IDE:
2) Choose Tools/Board/3D printer boards
3) Choose Tools/Board/Board Part Nr/Malyan M200 v1 or v2
4) .........

**Comment by xC0000005** [comment]
I’ve only been building in the arduino IDE, though most of my work lately has been on the lerdge boards, I did test both V1 and V2 after the boards were merged into main.

**Comment by thinkyhead** [comment]
STM32F1 build is broken due to some recent change in the platform. We've had to disable STM32 tests in `.travis.yml` here.

**Comment by thinkyhead** [comment]
@xC0000005 — I want to add a README.md for the M200 giving instructions on the build. Is Arduino IDE still the best option? If so, then I can just itemize the steps for Arduino. But if you have steps that work with the current PlatformIO platforms, I can add those instructions too. And, if the steps are not too complicated, then we may be able to add them to the Python build script to automate the process.
